### ISSUE TITLE
LDAP authentication requires too permissive LDAP ACLs

### ISSUE THREAD

**Issue by istvanszoke** [dpgaspar/Flask-AppBuilder#754]
Hello,

we are trying to configure an LDAP authentication in Superset (this application uses FAB) with the following config: (We use OpenLDAP)

```
AUTH_TYPE = AUTH_LDAP
AUTH_LDAP_SERVER = "ldap://ldap"
AUTH_LDAP_USE_TLS = False

AUTH_USER_REGISTRATION = True
AUTH_USER_REGISTRATION_ROLE = "Admin"

AUTH_LDAP_SEARCH = "ou=users,dc=example,dc=com"
AUTH_LDAP_BIND_USER = "CN=admin,DC=example,DC=com"
AUTH_LDAP_BIND_PASSWORD = "admin"

AUTH_LDAP_UID_FIELD = "uid"
AUTH_LDAP_FIRSTNAME_FIELD = "givenName"
AUTH_LDAP_LASTNAME_FIELD = "sn"
AUTH_LDAP_EMAIL_FIELD = "mail"
```

When FAB authenticates a user (i.e. szoke) it does the followings:
1. BIND with credentials given in (AUTH_LDAP_BIND_USER, AUTH_LDAP_BIND_PASSWORD)
2. SEARCH for the user that tries to login (i.e. szoke) along with the user's information 
3. BIND with user (i.e. szoke)
4. SEARCH for the user's information (firstname, lastname, email) (the reason for this request is unclear for us and this is the step that fails)

Theses steps are justified in the following log:

```
5b0fd5d8 conn=1011 fd=16 ACCEPT from IP=172.17.0.3:34152 (IP=0.0.0.0:389)
5b0fd5d8 conn=1011 op=0 BIND dn="cn=admin,dc=example,dc=com" method=128
5b0fd5d8 conn=1011 op=0 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
5b0fd5d8 conn=1011 op=0 RESULT tag=97 err=0 text=
5b0fd5d8 conn=1011 op=1 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(uid=szoke)"
5b0fd5d8 conn=1011 op=1 SRCH attr=givenName sn mail
5b0fd5d8 conn=1011 op=1 SEARCH RESULT tag=101 err=0 nentries=1 text=
5b0fd5d8 conn=1011 op=2 BIND anonymous mech=implicit ssf=0
5b0fd5d8 conn=1011 op=2 BIND dn="uid=szoke,ou=users,dc=example,dc=com" method=128
5b0fd5d8 conn=1011 op=2 BIND dn="uid=szoke,ou=users,dc=example,dc=com" mech=SIMPLE ssf=0
5b0fd5d8 conn=1011 op=2 RESULT tag=97 err=0 text=
5b0fd5d8 conn=1011 op=3 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(uid=szoke)"
5b0fd5d8 conn=1011 op=3 SRCH attr=givenName sn mail
5b0fd5d8 conn=1011 op=3 SEARCH RESULT tag=101 err=32 nentries=0 text=
```

According to the LOG the fourth step fails with error code 32, this means that the object could not be found. The reason for this is that the user does not have searching permission (in the LDAP) for the users LDAP entity that holds the actual user. (Remember, in the 3rd step we did a bind with the user) Giving permission to this would mean that a user can search for any other users in the system, which would cause security problems.

If we give a more permissive permission to the user the authentication works. 

If we look at a the beginning of the log, we see that the admin credential already searched for the user's information, so it is unclear why it is queried again in the fourth step. 

The question is why is there a 4th step? Are we missing something?

**Comment by dpgaspar** [comment]
Hi @istvanszoke,

Yes looks like you're right on self user registration mode and when the user is new, there is an extra ldap search that is being performed with the last bind, and at that time it's the user's bind

this is where step 4: https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/manager.py#L691

**Comment by alexandrezia** [comment]
Just create an ACL in ldap to allow user to see only his entry (self)

e.g. for OpenLDAP would be like this:
```
access to dn.subtree="ou=users,dc=foo,dc=bar"
        by self read
        by * none
```

**Comment by dpgaspar** [comment]
True, yet the correct way would be to use OpenLdap bind credentials on the second search also.

**Comment by istvanszoke** [comment]
@alexandrezia, your ACL is great. One additional is required which allows the users to search the users group... So to sum it up these ACLs are needed (note: the order is important)

```
to dn.exact="ou=users,dc=example,dc=com" by dn.one="ou=users,dc=example,dc=com" search
to dn.subtree="ou=users,dc=example,dc=com" by dn.exact="ou=users,dc=example,dc=com" manage by anonymous auth by self read by * none
```

@dpgaspar, since we found a work around, as far as I am concerned this issue can be closed.

Thanks for the help!

**Comment by dpgaspar** [comment]
Thank you @alexandrezia!

And @istvanszoke, I'll keep this open since I don't like the way the second search is done, but yes I will not consider this a urgent bug to solve.

**Comment by LLluma** [comment]
Hi all,
I faced the same problem. There is a fix #758.

**Comment by dpgaspar** [comment]
Great, looks good
